HACKING.adoc

```diff
@@ -302,6 +302,39 @@ If `boot/ocamlc` changes (e.g. because you ran `make bootstrap`), then
 the build will revert to the slower bytecode-compiled `ocamlc` until
 you do the above step again.
 
+=== Using merlin
+
+When building the compiler as described above, the cmi and cmt files that are
+produced are 99% of the time not readable by merlin, because it is usually
+updated to support new versions of the compiler after they get released.
+
+Therefore, to use merlin on the compiler, you want to build the compiler with an
+older version of itself. The easiest way to do this is to install the latest
+released version of the compiler, install dune, and run:
+
+----
+dune build @libs
+----
+
+which will do a bytecode build of all the distribution (without linking the
+executables), as well as generate a .merlin file.
+And merlin will be looking at the artefacts generated by dune (in `_build`),
+rather than try to open the ones produced by a Makefile build.
+
+Note however that for this to work properly, you need to have run the configure
+script, and the workspace needs to be clean, otherwise dune will complain that
+build artefacts are present among the sources (i.e. you need to `make clean`
+before calling dune).
+
+What this means is that whenever you change the interface of some compilation
+unit, and want merlin to be aware of the new interface, you will need to run
+`make clean && dune build @libs`.
+
+Finally, there will be times were the compiler simply cannot be built with an
+older version of itself. In which case, you'll either have to live without
+merlin, or first develop your patch on an older branch of the compiler before
+rebasing it (this can, however, introduce a substantial amount of extra work).
```

Comment:
This can introduce a substantial amount of extra work (when later rebasing to the current development version). But it also makes it easier to test the impact of your work on third-party code, by installing a local opam switch: opam packages tend to be compatible with release versions of the compiler, whereas many of them are incompatible with the current development version.

Reply:
I pushed a version that includes my suggestion.